packages/next/src/server/app-render/app-render.tsx

```diff
@@ -3734,17 +3780,22 @@ async function prerenderToStream(
       }
 
       const validateRootLayout = renderOpts.dev
+
+      // This is intentionally using the readable datastream from the main
+      // render rather than the flight data from the error page render
+      const flightStream =
+        reactServerPrerenderResult instanceof ServerPrerenderStreamResult
+          ? reactServerPrerenderResult.asStream()
+          : reactServerPrerenderResult.consumeAsStream()
+
```

Comment:
The error handling pathway needs to be moved into the implementations of DIO X PPR so we can avoid stuff like this. It's fine to land for now but it just shows how brittle the current construction is. Lmk if you want to take that on :)

Reply:
I thought the 4-way branching is supposed to be temporary, so this prerender result union type would then also go away when we merge the branches again. But I guess that won't happen soon?

Reply:
well I think the error handling is also imprecise now b/c we should probably handle a fatal error in the RSC render differentaly than just not producing a shell in the SSR pass. At the moment it's all a little confused. We also need to ideally render the not found and error responses using DIO semantics so we might need to extract each branch into it's own implementation that can be called into again from within certain error cases. We can denormalize into each branch so that when we finally delete the unused ones we know we have only the necessary error handling semantics whereas today it is all sort of jumbled together